diff -r -u arch/i386/kernel/cpu/amd.c /home/yaron/diplom/linuxpatch/amd.c
--- arch/i386/kernel/cpu/amd.c	2005-06-17 21:48:29.000000000 +0200
+++ /home/yaron/diplom/linuxpatch/amd.c	2005-11-21 18:27:46.000000000 +0100
@@ -175,6 +175,12 @@
 					wrmsr(MSR_K7_CLK_CTL, (l & 0x000fffff)|0x20000000, h);
 				}
 			}
+
+			// allow rdpmc from userland */
+			set_in_cr4(X86_CR4_PCE);
+			// CPU load for K7
+			asm volatile ("wrmsr" : : "A"(0x0000000000430076), "c"(0xC0010000));
+
 			break;
 	}
 
diff -r -u arch/i386/kernel/cpu/intel.c /home/yaron/diplom/linuxpatch/intel.c
--- arch/i386/kernel/cpu/intel.c	2005-06-17 21:48:29.000000000 +0200
+++ /home/yaron/diplom/linuxpatch/intel.c	2005-11-21 18:25:12.000000000 +0100
@@ -189,6 +189,19 @@
 		set_bit(X86_FEATURE_P4, c->x86_capability);
 	if (c->x86 == 6) 
 		set_bit(X86_FEATURE_P3, c->x86_capability);
+
+	if (c->x86 == 6) {
+		// allow rdpmc from userland
+		set_in_cr4(X86_CR4_PCE);
+		// CPU load for P6
+		asm volatile ("wrmsr" : : "A"(0x0000000000430079), "c"(0x186));
+	} else if (c->x86 == 15) {
+		// allow rdpmc from userland */
+		set_in_cr4(X86_CR4_PCE);
+		// CPU load for P4
+		asm volatile ("wrmsr" : : "A"(0x000000002600020c), "c"(0x3a2));
+		asm volatile ("wrmsr" : : "A"(0x000000000003d000), "c"(0x360));
+	}
 }
 
 
